---
title: "Introduction to DElegate"
author: "Christoph Hafemeister"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_float: yes
    code_folding: show
    highlight: pygments
    df_print: paged
  pdf_document:
    toc: yes
---

```{r, include = FALSE}
set.seed(11946992)
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

s <- readRDS(file = 'scratch/10k_pbmc_seurat.Rds')
```

```{r setup}
library(dplyr)
library(Seurat)
library(DElegate)
library(ggplot2)
library(ggrepel)
```

## Introduction

`DElegate` is an R package that allows bulk RNA-seq differential expression methods to be used with single-cell data. It is a light wrapper around 
[`DESeq2`](https://doi.org/doi:10.18129/B9.bioc.DESeq2), 
[`edgeR`](https://doi.org/doi:10.18129/B9.bioc.edgeR), and 
[`limma`](https://doi.org/doi:10.18129/B9.bioc.limma), similar to the [`Libra`](https://github.com/neurorestore/Libra) package. In contrast to `Libra`, `DElegate` focuses on a few DE methods and will assign cells to pseudo-replicates if no true replicates are available.

All `DElegate` functionality is contained in one function - `findDE()`. It has one mandatory input argument: `object`, which may be of class

* `Seurat` - the count matrix will be extracted from the `'RNA'` assay
* `SingleCellExperiment` - the count matrix will be extracted via `counts()`
* `dgCMatrix` - sparse matrix of the `Matrix` package
* `matrix`

To indicate the cell group memberships, you have several options, depending on input type:

* `Seurat` - in the object via `Idents(object)`, or use the `group_column` argument
* `SingleCellExperiment` - in the object via `colLables(object)`, or use the `group_column` argument
* `dgCMatrix`, or `matrix` - use the `meta_data` and `group_column` arguments

`DElegate` uses bulk RNA-seq DE methods and relies on replicates. If no true replicates are available, it assigns cells to pseudo-replicates. However, if replicates are available in the input, the `replicate_column` argument can be used to indicate where to find the labels.

To tell `findDE()` which cell groups to compare, use the `compare` argument. We provide several ways to set up the comparisons that will be tested:

* `'each_vs_rest'`, the default, does multiple comparisons, one per group vs all remaining cells
* `'all_vs_all'`, also does multiple comparisons, covering all group pairs
* a length one character vector, e.g. `'MONOCYTES'`, does one comparison between that group and the remaining cells
* a length two character vector, e.g. `c('T CELLS', 'B CELLS')`, does one comparison between those two groups
* a list of length two, e.g. `list(c('T CELLS', 'B CELLS'), c('MONOCYTES'))`, does one comparison after combining groups

Finally, there are currently three DE methods supported

* `'edger'` uses `edgeR::glmQLFit`
* `'deseq'` uses `DESeq2::DESeq(test = 'Wald')`
* `'limma'` uses `limma::eBayes(trend = TRUE, robust = TRUE)`


For complete details, consult the package documentation: `?DElegate::findDE`

In the examples below, we use a Seurat object `s` with cell type labels in the identity slot (`Idents(s)`).

```{r, fig.width=5, fig.height=3, out.width='50%'}
show(s)
table(Idents(s))
DimPlot(s)
```

## Testing for differential expression

### Each group vs rest (default)

```{r}
de_results <- findDE(object = s)
```

The output is one large data frame with the results for all three comparisons for all genes. The order of the output is by comparison, then by p-value, then by test statistic. Ordering can be turned off with `order_results = FALSE`
```{r}
head(de_results) %>%
  DT::datatable() %>%
  DT::formatSignif(columns = 2:8)
```

There are as many rows per comparison, as there were genes in the input
```{r}
table(paste(de_results$group1, de_results$group2, sep = ' vs '))
```

The output contains the following columns

* **feature** the gene (as given by rownames of input)
* **ave_expr** average expression of gene (renamed method specific values; edger: logCPM, deseq: baseMean, limma: Amean)
* **log_fc** log fold-change (renamed method specific values; edger: logFC, deseq: log2FoldChange, limma: LogFC)
* **stat** test statistic (renamed method specific values; edger: F, deseq: stat, limma: lods)
* **pvalue** test p-value
* **padj** adjusted p-value (FDR)
* **rate1** detection rate in group one (fraction of cells with non-zero counts)
* **rate2** detection rate in group two (fraction of cells with non-zero counts)
* **group1** comparison group one
* **group2** comparison group two

The log fold-change values are estimates of the form log2(group1/group2) - details will depend on the method chosen.

We can plot the log fold-change vs FDR for each comparison.

```{r, fig.width = 10, fig.height=3.6, out.width='100%'}
de_results$comp <- paste(de_results$group1, de_results$group2, sep = ' vs ')
de_results$comp <- factor(de_results$comp, levels = unique(de_results$comp))
ggplot(de_results, aes(log_fc, -log10(padj))) +
  geom_point(shape = 16, alpha = 0.2) +
  facet_wrap(~ comp, scales = 'free')
```

For a given comparison, the genes higher in the first group will be assigned positive fold-changes. Below we mark the top 5 per comparison, ranked by p-value (the default output of `findDE()`).

```{r, fig.width = 10, fig.height=3.6, out.width='100%'}
top_DE <- filter(de_results, log_fc > 0) %>%
      group_by(group1) %>%
      slice_head(n = 5)
ggplot(de_results, aes(log_fc, -log10(padj))) +
  geom_point(shape = 16, alpha = 0.2) +
  geom_point(data = top_DE, shape = 16, size = 2, color = 'deeppink') +
  geom_text_repel(data = top_DE, aes(label = feature)) +
  facet_wrap(~ comp, scales = 'free')
```

### One group vs another

To compare two specific groups, use a character vector of length two as the `'compare'` argument.

```{r}
de_results <- findDE(object = s, compare = c('T CELLS', 'B CELLS'))
```

```{r, fig.width = 5, fig.height=3.6, out.width='50%'}
top_DE <- filter(de_results, log_fc != 0) %>%
  group_by(sign(log_fc)) %>%
  slice_head(n = 5)
ggplot(de_results, aes(log_fc, -log10(padj))) +
  geom_point(shape = 16, alpha = 0.2) +
  geom_point(data = top_DE, shape = 16, size = 2, color = 'deeppink') +
  geom_text_repel(data = top_DE, aes(label = feature)) +
  ggtitle('T CELLS vs B CELLS')
```

### Non-Seurat input

An example of a count matrix and meta data as input.
```{r}
# extract the counts and meta data from our example Seurat object
counts <- s$RNA@counts
meta_data <- data.frame(celltype = Idents(s))
# use matrix and data frame as input to findDE
de_results <- findDE(object = counts, meta_data = meta_data, 
                     group_column = 'celltype', compare = c('T CELLS', 'B CELLS'))
head(de_results) %>%
  DT::datatable() %>%
  DT::formatSignif(columns = 2:8)
```

An example using a SingleCellExperiment as input.
```{r}
sce <- SingleCellExperiment::SingleCellExperiment(list(counts=counts))
SingleCellExperiment::colLabels(sce) <- meta_data$celltype
de_results <- findDE(object = sce, compare = c('T CELLS', 'B CELLS'))
head(de_results) %>%
  DT::datatable() %>%
  DT::formatSignif(columns = 2:8)
```


## Finding markers

To quickly find all marker features for a given object, `DElegate` provides the convenience function `FindAllMarkers2()` - a light wrapper around `findDE(compare = 'each_vs_rest')` that filters for positive fold-change and orders the results by p-value and test statistic. Note that by default this function applies a post-test filter that removes genes with low detection rate (`min_rate = 0.05`) and low log fold-change (`min_fc = 1`). 

```{r}
marker_results <- FindAllMarkers2(object = s)
filter(marker_results, feature_rank < 6) %>%
  DT::datatable() %>%
  DT::formatSignif(columns = 2:8)
```

## Parallelization

`DElegate` supports parallelization via the `future` package. 
For example, to use the multicore strategy with 12 workers you may call
`future::plan(strategy = 'future::multicore', workers = 12)` before DE testing.
See more details at the [`future` website](https://future.futureverse.org)

Note that every comparison is run single-threaded, but multiple comparisons will be done in parallel.

**Trouble shooting:** You may get an error regarding `future.globals.maxSize`, the maximum allowed total size of global variables. The default value is 500 MiB and may be too small. You may increase it, for example to 8GB, using `options(future.globals.maxSize = 8 * 10^9)`.

## Progress updates

For reporting progress updates, `DElegate` relies on the `progressr` package. By default no progress updates are rendered, but may be turned on in an R session: `progressr::handlers(global = TRUE)` and its default presentation modified (e.g. `progressr::handlers(progressr::handler_progress)`). 
See more details at the [`progressr` website](https://progressr.futureverse.org)

## Example: Seurat DE vignette

Here we perform the same DE tasks as in the [Seurat DE vignette](https://satijalab.org/seurat/articles/de_vignette.html).

```{r}
library(Seurat)
library(SeuratData)
pbmc <- LoadData("pbmc3k", type = "pbmc3k.final")
```

Find differentially expressed features between CD14+ and FCGR3A+ Monocytes (this is the alternative to `Seurat::FindMarkers(pbmc, ident.1 = "CD14+ Mono", ident.2 = "FCGR3A+ Mono")`)

```{r}
de_res <- findDE(object = pbmc, compare = c("CD14+ Mono", "FCGR3A+ Mono"))
head(de_res, 10) %>%
  DT::datatable() %>%
  DT::formatSignif(columns = 2:8)
```

We can omit the second part of the comparison argument to find CD14+ Mono markers relative to all other cells (alternative to `Seurat::FindMarkers(pbmc, ident.1 = "CD14+ Mono", ident.2 = NULL, only.pos = TRUE)`).

```{r}
de_res <- findDE(object = pbmc, compare = "CD14+ Mono")
# exclude negative log fold-change
filter(de_res, log_fc > 0) %>% head(10) %>%
  DT::datatable() %>%
  DT::formatSignif(columns = 2:8)
```

Note that we do not suggest prefiltering. A postfilter using log fold-change and/or detection rate may be appropriate instead.

Just as in the examples above, we can find markers for all groups at once with `FindAllMarkers2`. By default this wrapper around `findDE()` removes genes with detection rate < 0.05 (maximum across groups compared) and low log fold-change (< 1). This is an alternative to `Seurat::FindAllMarkers(pbmc)`.

```{r}
marker_results <- FindAllMarkers2(object = pbmc)
filter(marker_results, feature_rank < 6) %>%
  DT::datatable() %>%
  DT::formatSignif(columns = 2:8)
```


## Session info 

> **_NOTE:_** This document was generated with `DElegate` version `r packageVersion(pkg = 'DElegate')`

Session info
```{r}
sessionInfo()
```
